walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -40,19 +41,71 @@
     ArrayList<Event> sortedEvents = new ArrayList<>(events);
     Collections.sort(sortedEvents, Event.ORDER_BY_START);
 
+    Collection<Event> optionalEvents = new ArrayList<>();
+
+    ListIterator<Event> eventsIterator = sortedEvents.listIterator();
+    while (eventsIterator.hasNext()) {
+      Event event = eventsIterator.next();
+      
+      // Separates mandatory attendee events and optional attendee events into two lists and
+      // removes events that are attended by people not invited to the meeting. 
+      if (Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        // Events with only optional attendees are added to optionalEvents. 
+        if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+          optionalEvents.add(event);
+        }
+        // Events that don't have any mandatory attendees are removed from sortedEvents.
+        eventsIterator.remove();
+      }
+    }
+
+    Collection<TimeRange> mandatoryTimes = findAvailableTimes(sortedEvents, request);
+    Collection<TimeRange> optionalTimes = findAvailableTimes(optionalEvents, request);
+
+    // If there's no mandatory attendees, only consider optional attendees.
+    if (sortedEvents.size() == 0) {
+      return optionalTimes;
+    }
+
+    Collection<TimeRange> times = new ArrayList<>();
+
+    // Find common available times between mandatory and optional attendees. 
+    for (TimeRange mandatoryTime : mandatoryTimes) {
+      for (TimeRange optionalTime : optionalTimes) {
+        if (mandatoryTime.contains(optionalTime)) {
+          times.add(optionalTime);
+        }
+        else if (optionalTime.contains(mandatoryTime)) {
+          times.add(mandatoryTime);
+        }
+        else if (mandatoryTime.overlaps(optionalTime)) {
+          int start = mandatoryTime.start() > optionalTime.start() ? mandatoryTime.start() : optionalTime.start();
+          int end = mandatoryTime.end() < optionalTime.end() ? mandatoryTime.end() : optionalTime.end();
+
+          TimeRange time = TimeRange.fromStartEnd(start, end, false);
```

Comment:
If end==TimeRange.END_OF_DAY, then probably you will need 'true' here. See https://github.com/nif33/my-portfolio/pull/15#discussion_r442411386

Reply:
Actually I thought about this case and realized the only time when ```end == TimeRange.END_OF_DAY``` is when both ```mandatoryTime.end()``` and ```optionalTime.end()``` equal to ```TimeRange.END_OF_DAY```. In that case it wouldn't be an overlap, it'd be a case of contains, and the conditional blocks checking for that would catch it instead!

Reply:
aaah that makes sense thank you for explaining!